src/librustc_resolve/diagnostics.rs

```diff
@@ -444,6 +444,52 @@ impl SomeTrait for Foo { // ok!
 ```
 "##,
 
+E0406: r##"
+A function is referring to an associated type which hasn't been declared in
+the trait. Example of erroneous code:
+
+```compile_fail
+trait Foo {
+    type Bar;
+
+    // error: function signature contains a reference to `Self::Baz`, but
+    //        `Baz` hasn't been declared as an associated type of the trait
+    fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
+}
+```
+
+The use of 'associated types' allows us declare variables which might be used
+in any of the function signatures in a particular trait. This helps in
+eliminating redundancies and improves readability of code. It works by using
```

Comment:
"helps eliminate redundancies and improve readability" would have more [parallelism](https://en.wikipedia.org/wiki/Parallelism_%28grammar%29) (_eliminate_ and _improve_ being [bare infinitives](https://en.wikipedia.org/wiki/Uses_of_English_verb_forms#Bare_infinitive) complementing _helps_).